File: boot-web-security-parent/boot-ch11-business-logic-server/src/main/java/com/yz/config/SecurityConfigure.java
package com.yz.config;

import com.yz.authentication.provider.OtpAuthenticationProvider;
import com.yz.authentication.provider.UsernamePasswordAuthenticationProvider;
import com.yz.filter.JwtAuthenticationFilter;
import com.yz.filter.PreAuthFilter;
import com.yz.proxy.AuthenticationServerProxy;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;

/**
 * @author andrew
 * @date 2020-11-04
 */
@Configuration
public class SecurityConfigure extends WebSecurityConfigurerAdapter {

    @Value("${sign.key}")
    private String signKey;
    private final AuthenticationServerProxy proxy;

    public SecurityConfigure(AuthenticationServerProxy proxy) {
        this.proxy = proxy;
    }

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.csrf().disable();
        http.addFilterAt(new PreAuthFilter(this.signKey, authenticationManagerBean()),
                BasicAuthenticationFilter.class)
                .addFilterAfter(new JwtAuthenticationFilter(this.signKey), BasicAuthenticationFilter.class);

        http.authorizeRequests().anyRequest().authenticated();
    }

    @Override
    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
        // register providers
        auth.authenticationProvider(new UsernamePasswordAuthenticationProvider(this.proxy))
                .authenticationProvider(new OtpAuthenticationProvider(this.proxy));
    }

    @Bean
    @Override
    public AuthenticationManager authenticationManagerBean() throws Exception {
        return super.authenticationManagerBean();
    }
}

File: cloud-oauth2-server/src/main/kotlin/com/yz/config/SecurityConfigure.kt
package com.yz.config

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.config.BeanIds
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.provisioning.InMemoryUserDetailsManager

/**
 * web security 配置
 * @author andrew
 * @date 2020-09-18
 */
@EnableWebSecurity
class SecurityConfigure : WebSecurityConfigurerAdapter() {

    override fun configure(http: HttpSecurity) {
        http
                .authorizeRequests()
                .antMatchers("/oauth2/keys").permitAll()
                // 其他接口请求需要登陆认证
                .anyRequest().authenticated()
                .and()
                .formLogin()
    }

    @Bean
    fun passwordEncoder() : PasswordEncoder = BCryptPasswordEncoder()

    @Bean
    fun users(): UserDetailsService {
        val guest = User.withUsername("guest").password(passwordEncoder().encode("guest")).roles("GUEST").build()
        val admin = User.withUsername("admin").password(passwordEncoder().encode("admin")).roles("GUEST", "ADMIN").build()
        return with(InMemoryUserDetailsManager()) {
            this.createUser(guest)
            this.createUser(admin)
            this
        }
    }

    @Bean(name = [BeanIds.AUTHENTICATION_MANAGER])
    override fun authenticationManagerBean(): AuthenticationManager {
        return super.authenticationManagerBean()
    }

}